src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/TerminalPane.java

```diff
@@ -15,227 +15,201 @@
    {
-      secureInput_.secureString(event.getData(), new CommandWithArg<String>() 
+      if (secureInput_ == null)
       {
-         @Override
-         public void execute(String arg) // success
-         {
-            consoleProcess_.writeStandardInput(
-               ShellInput.create(arg,  true /* echo input*/), 
-               new VoidServerRequestCallback() {
-                  @Override
-                  public void onError(ServerError error)
-                  {
-                     xterm_.writeln(error.getUserMessage());
-                  }
-               });
-          }
-      },
-      new CommandWithArg<String>()
-      {
-         @Override
-         public void execute(String errorMessage) // failure
-         {
-            xterm_.writeln(errorMessage); 
-         }
-      });
+         secureInput_ = new ShellSecureInput();  
+      }
+      
+      TerminalSession newSession = new TerminalSession(secureInput_);
+      
+      addHandlerRegistration(newSession, newSession.addTerminalSessionStartedHandler(this));
+      addHandlerRegistration(newSession, newSession.addTerminalSessionStoppedHandler(this));
+      
+      newSession.connect();
    }
+
+   @Override
+   public void onTerminalSessionStarted(TerminalSessionStartedEvent event)
+   {
+      terminalSessionsPanel_.add(event.getTerminalWidget());
+      terminalSessionsPanel_.showWidget(event.getTerminalWidget());
+      
+      // TODO (gary) update dropdown
+   } 
    
    @Override
-   protected Toolbar createMainToolbar()
+   public void onTerminalSessionStopped(TerminalSessionStoppedEvent event)
    {
-      Toolbar toolbar = new Toolbar();
+      Widget currentTerminal = event.getTerminalWidget();
+      unregisterHandlers(currentTerminal);
+      int currentIndex = terminalSessionsPanel_.getWidgetIndex(currentTerminal);
+      if (currentIndex > 0)
+      {
+         terminalSessionsPanel_.showWidget(currentIndex - 1);
+         // TODO (gary) set focus on the widget
+      }
+      terminalSessionsPanel_.remove(currentTerminal);
+       
+      // TODO (gary) update dropdown
+   }
 
-      activeTerminalToolbarButton_ = new TerminalPopupMenu(sessionInfo_,
-                                                           commands_);
-      
-      toolbar.addLeftWidget(activeTerminalToolbarButton_.getToolbarButton());
+   public int getTerminalCount()
+   {
+      return terminalSessionsPanel_.getWidgetCount();
+   }
 
-      return toolbar;
+   @Override
+   public void onUnload()
+   {
+      super.onUnload();
+      unregisterAllHandlers();
+   }
+
+   /**
+    * Track a handler registration with a specific widget, so we can remove
+    * handlers for that specific widget when it is removed.
+    * @param widget Widget we are registering with
+    * @param reg Handler we are tracking
+    */
+   protected void addHandlerRegistration(Widget widget, HandlerRegistration reg)
+   {
+      HandlerRegistrations registrations = widgetToRegistrations_.get(widget);
```

Comment:
Would it be simpler/safer to store the handler registrations in the widget, and remove them when the widget is detached?